dojo/pipeline.py

```diff
@@ -69,6 +75,63 @@ def modify_permissions(backend, uid, user=None, social=None, *args, **kwargs):
                 user.is_staff = False
 
 
+def update_azure_groups(backend, uid, user=None, social=None, *args, **kwargs):
+    if settings.AZUREAD_TENANT_OAUTH2_ENABLED and settings.AZUREAD_TENANT_OAUTH2_GET_GROUPS and isinstance(backend, AzureADTenantOAuth2):
+        soc = user.social_auth.get()
+        token = soc.extra_data['access_token']
+        group_names = []
+        if 'groups' not in kwargs['response'] or kwargs['response']['groups'] == "":
+            logger.warn("No groups in response. Stopping to update groups of user based on azureAD")
+            return
+        group_IDs = kwargs['response']['groups']
+        for group_from_response in group_IDs:
+            logger.debug("Analysing Group_ID " + group_from_response)
+            request_headers = {'Authorization': 'Bearer ' + token}
+            try:
+                if is_group_id(group_from_response):
+                    logger.debug("detected " + group_from_response + " as groupID and will fetch the displayName from microsoft graph")
+                    group_name_request = requests.get((str(soc.extra_data['resource']) + '/v1.0/groups/' + str(group_from_response) + '?$select=displayName'), headers=request_headers)
+                    group_name_request_json = group_name_request.json()
+                    group_name = group_name_request_json['displayName']
+                else:
+                    logger.debug("detected " + group_from_response + " as group name and will not call microsoft graph")
+                    group_name = group_from_response
+
+                group, created_group = Dojo_Group.objects.get_or_create(name=group_name, is_azure=True)
+                if settings.AZUREAD_TENANT_OAUTH2_GROUPS_FILTER == "" or re.search(settings.AZUREAD_TENANT_OAUTH2_GROUPS_FILTER, group_name):
+                    group_names.append(group_name)
+                else:
+                    logger.debug("Skipping group " + group_name + " due to AZUREAD_TENANT_OAUTH2_GROUPS_FILTER " + settings.AZUREAD_TENANT_OAUTH2_GROUPS_FILTER)
+                    continue
+
+                group_member, is_member_created = Dojo_Group_Member.objects.get_or_create(group=group, user=user, defaults={
+                    'role': Role.objects.get(id=Roles.Writer)})
+
+                if settings.AZUREAD_TENANT_OAUTH2_CLEANUP_GROUPS:
+                    cleanup_old_azureAD_groups_for_user(user, group_names)
+            except:
+                logger.error("Could not call microsoft graph API or save groups to member")
+                traceback.print_exc()
```

Comment:
If there is no permission to query the graph API, this will lead to lots and lots of exceptions/errors in the logs?

Reply:
Correct. In case there is no permission, you woudn't activate `AZUREAD_TENANT_OAUTH2_ENABLED`.

Reply:
Is it  a valid usecase to have
a) azure ad oauth enabled
b) return group ids in token
c) do not give permission to graph api to retrieve groups
I think so, but some of the things I mentioned in comments yesterday could be future PRs, same here.